File: onecard-client/lib/utils/cardUtils.ts
import { Player } from '@/types/gamePlayer';
import { Direction, GameState } from '@/types/gameState';
import {
	PokerCardProps,
	PokerCardPropsWithId,
	RankValue,
	SuitsValue,
} from '@/types/pokerCard';

export const createDeck = (includeJokers: boolean): PokerCardPropsWithId[] => {
	const suits: SuitsValue[] = ['hearts', 'diamonds', 'clubs', 'spades'];
	const ranks: RankValue[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];

	let deck: PokerCardPropsWithId[] = suits.flatMap(suit =>
		ranks.map(rank => ({
			id: crypto.randomUUID(),
			suit,
			rank,
			isJoker: false,
			isFlipped: true,
			draggable: false,
		})),
	);

	if (includeJokers) {
		deck.push(
			{
				id: crypto.randomUUID(),
				isJoker: true,
				isFlipped: true,
				draggable: false,
			},
			{
				id: crypto.randomUUID(),
				isJoker: true,
				isFlipped: true,
				draggable: false,
			},
		);
	}

	return deck;
};

export const shuffleDeck = (
	deck: PokerCardPropsWithId[],
): PokerCardPropsWithId[] => {
	const shuffled = [...deck];
	for (let i = shuffled.length - 1; i > 0; i--) {
		const j = Math.floor(Math.random() * (i + 1));
		[shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
	}
	return shuffled;
};

export const refillDeck = (
	currentDeck: PokerCardPropsWithId[],
	discardPile: PokerCardPropsWithId[],
): {
	newDeck: PokerCardPropsWithId[];
	newDiscardPile: PokerCardPropsWithId[];
} => {
	if (discardPile.length === 0) {
		return { newDeck: currentDeck, newDiscardPile: [] };
	}

	const cardsToShuffle = [...currentDeck, ...discardPile.slice(1)];
	const shuffledDeck = shuffleDeck(cardsToShuffle);
	const newDiscardPile = [discardPile[0]];

	return {
		newDeck: shuffledDeck,
		newDiscardPile: newDiscardPile,
	};
};

export const dealCards = (
	players: Player[],
	deck: PokerCardPropsWithId[],
	initHandSize: number,
): { updatedPlayers: Player[]; updatedDeck: PokerCardPropsWithId[] } => {
	const updatedPlayers = players.map((player, idx) => ({
		...player,
		hand: deck.slice(idx * initHandSize, idx * initHandSize + initHandSize),
	}));

	const updatedDeck = deck.slice(players.length * initHandSize);

	return { updatedPlayers, updatedDeck };
};

export const isAbleToBlock = (
	playedCard: PokerCardProps,
	topCard: PokerCardProps,
	damage: number,
): boolean => {
	if (topCard.rank === 2) {
		return (
			playedCard.rank === 2 ||
			(playedCard.suit === topCard.suit && playedCard.rank === 1)
		);
	} else if (topCard.rank === 1) {
		return playedCard.rank === 1;
	}
	return playedCard.isJoker;
};

export const attackValue = (card: PokerCardProps): number => {
	if (card.rank === 2) return 2; // 2 는 2장
	if (card.rank === 1) return 5; // A 는 5장
	if (card.isJoker) return 7; // 조커는 7장
	return 0;
};

export const changeDirection = (
	card: PokerCardProps,
	curDirection: Direction,
): Direction => {
	if (card.rank === 12)
		return curDirection === 'clockwise' ? 'counterclockwise' : 'clockwise'; // Q 에서 반전
	return curDirection;
};

export const turnSpecialEffect = (
	card: PokerCardProps,
	state: GameState,
): number => {
	if (card.rank === 11) return getNextPlayerIndex(state);
	else if (card.rank === 13) return getPrevPlayerIndex(state);
	return state.currentPlayerIndex;
};

// 카드의 유효성 검사 함수
export const isValidPlay = (
	playedCard: PokerCardProps,
	topCard: PokerCardProps,
	damage: number,
): boolean => {
	if (playedCard.isJoker) return true;
	if (damage > 0) return isAbleToBlock(playedCard, topCard, damage);
	if (!playedCard.rank || !playedCard.suit || !topCard.rank || !topCard.suit)
		return false;
	return playedCard.rank === topCard.rank || playedCard.suit === topCard.suit;
};

export const checkWinner = (players: Player[]): Player | null => {
	const winner = players.find(player => player.hand.length === 0);
	return winner || null;
};

export const getNextPlayerIndex = (state: GameState): number => {
	const playerCount = state.players.length;
	if (state.direction === 'clockwise') {
		return (state.currentPlayerIndex + 1) % playerCount;
	} else {
		return (state.currentPlayerIndex - 1 + playerCount) % playerCount;
	}
};

export const getPrevPlayerIndex = (state: GameState): number => {
	const playerCount = state.players.length;
	if (state.direction === 'clockwise') {
		return (state.currentPlayerIndex - 1 + playerCount) % playerCount;
	} else {
		return (state.currentPlayerIndex + 1) % playerCount;
	}
};

File: onecard-server/src/onecard_api/domain/card_utils.py
from __future__ import annotations

import random
import uuid
from typing import Iterable

from .types import Direction, GameState, Player, PokerCard, RankValue, SuitValue

SUIT_VALUES: tuple[SuitValue, ...] = ("hearts", "diamonds", "clubs", "spades")
RANK_VALUES: tuple[RankValue, ...] = tuple(range(1, 14))  # 1~13


def create_deck(include_jokers: bool) -> list[PokerCard]:
    deck: list[PokerCard] = [
        {
            "id": str(uuid.uuid4()),
            "suit": suit,
            "rank": rank,
            "isJoker": False,
            "isFlipped": True,
            "draggable": False,
        }
        for suit in SUIT_VALUES
        for rank in RANK_VALUES
    ]

    if include_jokers:
        deck.append(
            {
                "id": str(uuid.uuid4()),
                "isJoker": True,
                "isFlipped": True,
                "draggable": False,
            }
        )
        deck.append(
            {
                "id": str(uuid.uuid4()),
                "isJoker": True,
                "isFlipped": True,
                "draggable": False,
            }
        )

    return deck


def shuffle_deck(deck: Iterable[PokerCard]) -> list[PokerCard]:
    shuffled = list(deck)
    random.shuffle(shuffled)
    return shuffled


def refill_deck(
    current_deck: list[PokerCard], discard_pile: list[PokerCard]
) -> dict[str, list[PokerCard]]:
    if not discard_pile:
        return {"new_deck": list(current_deck), "new_discard_pile": []}

    cards_to_shuffle = [*current_deck, *discard_pile[1:]]
    shuffled_deck = shuffle_deck(cards_to_shuffle)
    new_discard_pile = [discard_pile[0]]
    return {"new_deck": shuffled_deck, "new_discard_pile": new_discard_pile}


def deal_cards(
    players: list[Player], deck: list[PokerCard], init_hand_size: int
) -> tuple[list[Player], list[PokerCard]]:
    updated_players: list[Player] = []
    for idx, player in enumerate(players):
        start = idx * init_hand_size
        end = start + init_hand_size
        updated_players.append(
            {**player, "hand": list(deck[start:end])}  # copy slice
        )

    updated_deck = list(deck[len(players) * init_hand_size :])
    return updated_players, updated_deck


def is_able_to_block(played_card: PokerCard, top_card: PokerCard) -> bool:
    if top_card.get("rank") == 2:
        return played_card.get("rank") == 2 or (
            played_card.get("suit") == top_card.get("suit")
            and played_card.get("rank") == 1
        )
    if top_card.get("rank") == 1:
        return played_card.get("rank") == 1
    return bool(played_card.get("isJoker"))


def attack_value(card: PokerCard) -> int:
    if card.get("rank") == 2:
        return 2
    if card.get("rank") == 1:
        return 5
    if card.get("isJoker"):
        return 7
    return 0


def change_direction(card: PokerCard, cur_direction: Direction) -> Direction:
    if card.get("rank") == 12:
        return "counterclockwise" if cur_direction == "clockwise" else "clockwise"
    return cur_direction


def turn_special_effect(card: PokerCard, state: GameState) -> int:
    if card.get("rank") == 11:
        return get_next_player_index(state)
    if card.get("rank") == 13:
        return get_prev_player_index(state)
    return state["currentPlayerIndex"]


def is_valid_play(
    played_card: PokerCard, top_card: PokerCard, damage: int
) -> bool:
    if played_card.get("isJoker"):
        return True
    if damage > 0:
        return is_able_to_block(played_card, top_card)
    if top_card.get("isJoker"):
        return True

    if played_card.get("rank") is None or played_card.get("suit") is None:
        return False
    if top_card.get("rank") is None or top_card.get("suit") is None:
        return False

    return played_card.get("rank") == top_card.get("rank") or played_card.get(
        "suit"
    ) == top_card.get("suit")


def check_winner(players: Iterable[Player]) -> Player | None:
    for player in players:
        if len(player.get("hand", [])) == 0:
            return player
    return None


def get_next_player_index(state: GameState) -> int:
    player_count = len(state["players"])
    if state["direction"] == "clockwise":
        return (state["currentPlayerIndex"] + 1) % player_count
    return (state["currentPlayerIndex"] - 1 + player_count) % player_count


def get_prev_player_index(state: GameState) -> int:
    player_count = len(state["players"])
    if state["direction"] == "clockwise":
        return (state["currentPlayerIndex"] - 1 + player_count) % player_count
    return (state["currentPlayerIndex"] + 1) % player_count
